mesonbuild/dependencies/base.py

```diff
@@ -579,38 +578,50 @@ def _set_libs(self):
         libcmd = [self.name, '--libs']
         if self.static:
             libcmd.append('--static')
-            # Force pkg-config to output -L fields even if they are system
-            # paths so we can do manual searching with cc.find_library() later.
-            env = os.environ.copy()
-            env['PKG_CONFIG_ALLOW_SYSTEM_LIBS'] = '1'
+        # Force pkg-config to output -L fields even if they are system
+        # paths so we can do manual searching with cc.find_library() later.
+        env = os.environ.copy()
+        env['PKG_CONFIG_ALLOW_SYSTEM_LIBS'] = '1'
         ret, out = self._call_pkgbin(libcmd, env=env)
         if ret != 0:
             raise DependencyException('Could not generate libs for %s:\n\n%s' %
                                       (self.name, out))
-        self.link_args = []
+        # These libraries should be safe to de-dup
+        link_args = OrderedSet()
         libpaths = []
-        static_libs_notfound = []
+        libs_notfound = []
+        libtype = 'static' if self.static else 'default'
+        # We always look for the file ourselves instead of depending on the
+        # compiler to find it with -lfoo or foo.lib (if possible) because:
+        # 1. We want to be able to select static or shared
+        # 2. We need the full path of the library to calculate RPATH values
+        #
+        # Libraries that are provided by the toolchain or are not found by
+        # find_library() will be added with -L -l pairs.
         for lib in self._convert_mingw_paths(shlex.split(out)):
-            # If we want to use only static libraries, we have to look for the
-            # file ourselves instead of depending on the compiler to find it
-            # with -lfoo or foo.lib. However, we can only do this if we already
-            # have some library paths gathered.
-            if self.static:
-                if lib.startswith('-L'):
-                    libpaths.append(lib[2:])
-                    continue
-                # FIXME: try to handle .la files in static mode too?
-                elif lib.startswith('-l'):
-                    args = self.compiler.find_library(lib[2:], self.env, libpaths, libtype='static')
-                    if not args or len(args) < 1:
-                        if lib in static_libs_notfound:
-                            continue
+            if lib.startswith('-L'):
+                libpaths.append(lib[2:])
+                continue
+            elif lib.startswith('-l'):
+                args = self.compiler.find_library(lib[2:], self.env, libpaths, libtype)
```

Comment:
My code had an error message for `self.compiler is None`. I don't remember the exact code paths that triggered that but i think we need this here too.

Reply:
Yes, I noticed that but I didn't want to add untested or potentially dead code. So I thought I'd skip it and see what breaks so we know what that case is.

Reply:
I'm pretty sure it's not dead. And i don't think leaving preventable stack traces around for unsuspecting users is a good approach.

Reply:
I don't like preemptively adding checks for cases without understanding what the consequences of those cases are, since silently taking a bad decision is worse. I'll add an assert telling the user to file a bug.

Reply:
can you try in a java only (as in meson languages) project? I know that that is nonsensical, but maybe it's an easy way to reproduce it.

Reply:
I read the code again, and this can happen if the project doesn't have a clike language in the list of languages in `project()` or `add_languages()`: `rust`, `csharp`, `python`. So I re-added the check.